### ISSUE TITLE
Single instance of Kroxylicious with port-per-broker exposition not resilient to restart/crash

### ISSUE THREAD

**Issue by robobario** [kroxylicious/kroxylicious#416]
**Describe the bug**

Given we have a single Kroxylicious instance proxying a 3-node broker and it has a single Producer connected and sending data to all brokers. If Kroxylicious it is killed (by deleting the pod), then when it comes back up the existing Kafka producer is left in a bad state, unable to connect to the brokers.

Tested by deleting the kroxylicious pod while running the producer performance test bundled with kafka (uses the high level java Producer). The producer performance test spews messages like:
```
[2023-06-27 22:05:10,277] WARN [Producer clientId=perf-producer-client] Connection to node 2 (kroxylicious-service/10.96.20.28:9295) could not be established. Broker may not be available. (org.apache.kafka.clients.NetworkClient)
[2023-06-27 22:05:11,047] WARN [Producer clientId=perf-producer-client] Connection to node 1 (kroxylicious-service/10.96.20.28:9294) could not be established. Broker may not be available. (org.apache.kafka.clients.NetworkClient)
[2023-06-27 22:05:11,247] WARN [Producer clientId=perf-producer-client] Connection to node 0 (kroxylicious-service/10.96.20.28:9293) could not be established. Broker may not be available. (org.apache.kafka.clients.NetworkClient)
[2023-06-27 22:05:11,348] WARN [Producer clientId=perf-producer-client] Connection to node 2 (kroxylicious-service/10.96.20.28:9295) could not be established. Broker may not be available. (org.apache.kafka.clients.NetworkClient)
```
I left it for 10 minutes before killing the test, was wondering if it would eventually refresh the metadata using the bootstrap address but it appears not to.

**To Reproduce**
1. Start kroxylicious on minikube `minikube delete && QUAY_ORG=${your_quay_id_here} ./scripts/run-with-strimzi.sh`
2. Run command '`kubectl -n kafka run kafka-producer -ti --image=quay.io/strimzi/kafka:0.35.0-kafka-3.4.0 --rm=true --restart=Never -- bin/kafka-producer-perf-test.sh --topic my-topic --producer-props bootstrap.servers=kroxylicious-service:9292 --num-records 20000000 --record-size 100 --throughput -1`
3. Wait for the test to report throughput a couple of times and then delete the pod like `kubectl delete pod -nkafka kroxylicious-proxy-7df874b999-nrtrg`, fill in your pod name here.

**Expected behavior**
Recovery of the Producer within some time period.

**Additional context**
The series of events is likely something like:

1. producer connects to bootstrap port (bound on startup by Kroxylicious) and requests metadata
2. Kroxylicious intercepts metadata response and binds a port per broker. Responds, telling producer how to address those ports.
3. Producer connects to the broker ports and gets producing
4. Kroxylicious restarts, only the bootstrap port is bound
5. Producer keeps trying to reconnect to broker ports, metadata is not requested again on the bootstrap port so we can't drive port-per-broker binding.

**Comment by robobario** [comment]
Some brainstorming of options (may be rubbish, off the top of my head :grin: ):
1. Give Kroxylicious power request metadata itself with an admin client.
2. Make kroxy stateful so it can get at the last metadata for a target cluster somehow, probably requires some external storage.
3. Bind all the ports on startup, and make them do something on the initial connection like connect to the target bootstrap address, request metadata, wait for the response and then close the channel. Then on subsequent calls they can proxy to the correct broker.
4. Offer a more clunky, non-dynamic style. Pre-bind a port per broker and use a pattern to construct the target address to connect to. So if I get a connection on 9193, I know it's broker 1 and then use pattern `${nodeId}-real-broker.internal.kafka:9192` to connect.

**Comment by robobario** [comment]
Created a hacky implementation of option 3 to see if it could work (prebind the ports, request metadata if one of the broker-ports is hit, trigger reconciliation and close connection). https://github.com/robobario/kproxy/tree/spike-prebind

It recovers in this simple test, but I don't know if it would work with SASL connections (can we request broker metadata before authentication? need to understand the flow there)

**Comment by showuon** [comment]
> It recovers in this simple test, but I don't know if it would work with SASL connections (can we request broker metadata before authentication? need to understand the flow there)

SASL will request metadata before doing authentication, so it should be fine. BUT, SSL_SASL will need to do SSL handshake -> metadata request -> SASL authentication, so, not sure if it works in this case

**Comment by showuon** [comment]
OK, checked the code, and I was wrong. It turns out the request before SASL authentication is API_VERSION, not METADATA we wants. So, that means all metadata request will go after authentication. FYI
https://github.com/apache/kafka/blob/trunk/core/src/main/scala/kafka/server/KafkaApis.scala#L1860C1-L1860C1

**Comment by robobario** [comment]
Thanks @showuon !

**Comment by showuon** [comment]
> Some brainstorming of options (may be rubbish, off the top of my head 😁 ):
> 
>     1. Give Kroxylicious power to request metadata itself with an admin client. More configuration for the user.
> 
>     2. Make kroxy stateful so it can get at the last metadata for a target cluster somehow, probably requires some external storage.
> 
>     3. Bind all the ports on startup, and make them do something on the initial connection like connect to the target bootstrap address, request metadata, wait for the response and then close the channel. Then on subsequent calls they can proxy to the correct broker.
> 
>     4. Offer a more clunky, non-dynamic style. Pre-bind a port per broker and use a pattern to construct the target address to connect to. So if I get a connection on 9193, I know it's broker 1 and then use pattern `${nodeId}-real-broker.internal.kafka:9192` to connect.

I'd vote for option (2), which should be the easiest solution. We need to make sure we flush the metadata into the persistent storage before responding to clients. We can also store some more info when we support kroxy cluster in the future.

**Comment by robobario** [comment]
We had some discussion not too long ago about using a compacted kafka topic for state, so I guess in that case we'd have to configure Kroxylicious with properties to consume/produce from a kafka (the user could point it at their existing cluster).

**Comment by robobario** [comment]
Need to look at what happens in the SNI case, the port will be bound but I guess that it won't know the address to forward to (vaguely recall code that would use the upstream bootstrap address?)

**Comment by SamBarker** [comment]
> Need to look at what happens in the SNI case, the port will be bound but I guess that it won't know the address to forward to (vaguely recall code that would use the upstream bootstrap address?)

Yeah I think SNI does something similar but can't point to the code quickly. 

Personally I'm leaning towards 3. 

Option 2 seems like a useful fall back but I'd like to avoid the additional config of managing users/acls/service accounts if possible.

**Comment by robobario** [comment]
For option 3, I think we could work around the SASL auth by only intercepting the non-SASL-related messages (and non-ApiVersions). So we would discard and request metadata after SASL auth has completed.

**Comment by k-wall** [comment]
I think 3 is the way to go too.  In the PortPerBroker case, it does have the disadvantage that Kroxylicious will have ports open needlessly in the case where the cluster is smaller than the actual cluster.  I suppose this doesn't matter too much.

I have a feeling  we might need to solve #351 before the the solution will work well in the case where Kroxylicious is restarted whilst clients are connnected.

**Comment by k-wall** [comment]
I'd like to work on this issue.

**Comment by robobario** [comment]
I gave it a [spike](https://github.com/robobario/kproxy/tree/spike-prebind) when I created the issue if it's any help @k-wall, code is very scrappy but maybe the concept could be pulled out. The implementation includes:

1. Add some state to `VirtualClusterBrokerBinding` so that the binding knows if in this inital state. 
2. Call the endpoint reconciler to do the binding after the bootstrap port is bound, for all brokers up to maxNumBrokers. As part of this call we have to tell the reconciler that these are 'initial' bindings so the state can be set on `VirtualClusterBrokerBinding`.
3. When we are setting up the Filters for a binding, if it's in the initial state, instead of adding the `BrokerAddressFilter` we add a new filter that knows how to forward until we can send a metadata request to the broker.
4. When we intercept the next message, we instead send a metadata request and drive endpoint reconciliation on the response, then close the connection (needs a method on the `KrpcFilterContext`).

Not sure what the SNI side would look like with this implementation, you could add in some amount of `VirtualClusterBrokerBinding` the same way, maybe with a high default `maxNumBrokers` so most users would never have to configure it, like 128. Or maybe we could do it with regexes, like if we don't have a binding for `01-broker.kafka.internal` and the pattern is `${brokerId}-broker.kafka.internal` we could locate the matching virtual cluster and return a new `VirtualClusterBrokerBinding` referencing the upstream bootstrap and in the 'initial' state.
